File: kakao.app/js/ked/view/base/crumbs.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.prototype.hasOwnProperty(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, min: function () { var m = Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.min.apply(_k_.min,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n < m ? n : m}}}; return m }, lpad: function (l,s='',c=' ') {s=String(s); while(s.length<l){s=c+s} return s}, trim: function (s,c=' ') {return _k_.ltrim(_k_.rtrim(s,c),c)}, ltrim: function (s,c=' ') { while (_k_.in(s[0],c)) { s = s.slice(1) } return s}, rtrim: function (s,c=' ') {while (_k_.in(s.slice(-1)[0],c)) { s = s.slice(0, s.length - 1) } return s}}

var crumbs

import kxk from "../../../kxk.js"
let slash = kxk.slash
let post = kxk.post

import color from "../../theme/color.js"
import theme from "../../theme/theme.js"

import view from "./view.js"


crumbs = (function ()
{
    _k_.extend(crumbs, view)
    function crumbs (screen, name)
    {
        this["onMouse"] = this["onMouse"].bind(this)
        this["show"] = this["show"].bind(this)
        crumbs.__super__.constructor.call(this,screen,name)
        this.pointerType = 'pointer'
        this.rounded = ''
        this.setColor('bg',theme.crumbs.bg)
        this.setColor('fg',theme.crumbs.fg)
        this.setColor('empty_left','#000')
        this.setColor('empty_right','#000')
    }

    crumbs.prototype["setColor"] = function (key, color)
    {
        switch (key)
        {
            case 'empty':
                this.setColor('empty_left',color)
                this.setColor('empty_right',color)
                return

        }

        return crumbs.__super__.setColor.call(this,key,color)
    }

    crumbs.prototype["layout"] = function (x, y, w, h)
    {
        crumbs.__super__.layout.call(this,x,y,w,h)
    
        return this.adjustText()
    }

    crumbs.prototype["draw"] = function ()
    {
        var bg, ch, colors, fg, i, si, x

        if (this.hidden())
        {
            return
        }
        if (_k_.empty(this.rounded))
        {
            return
        }
        colors = []
        for (var _a_ = i = 0, _b_ = this.split.length; (_a_ <= _b_ ? i < this.split.length : i > this.split.length); (_a_ <= _b_ ? ++i : --i))
        {
            if (i === this.hoverIndex)
            {
                colors.push(color.values('#44a'))
            }
            else
            {
                colors.push(color.darken(this.color.bg,0.4 + 0.6 * (i + 1) / this.split.length))
            }
        }
        for (var _c_ = x = 0, _d_ = this.rounded.length; (_c_ <= _d_ ? x < this.rounded.length : x > this.rounded.length); (_c_ <= _d_ ? ++x : --x))
        {
            si = this.splitIndexAtCol(x)
            bg = colors[si]
            ch = this.rounded[x]
            if (_k_.in(ch,''))
            {
                fg = bg
                bg = (si > 0 ? colors[si - 1] : this.color.empty_left)
                if (x === this.rounded.length - 1)
                {
                    bg = this.color.empty_right
                }
                this.cells.set(x,0,ch,fg,bg)
            }
            else
            {
                if (si === this.hoverIndex)
                {
                    fg = color.adjustForBackground(this.color.fg,bg)
                }
                else if (si < this.split.length - 1)
                {
                    fg = color.darken(this.color.fg,_k_.min(1,(si + 3) / this.split.length))
                }
                else
                {
                    fg = theme.syntax.dir_leaf
                }
                this.cells.set(x,0,ch,fg,bg)
            }
        }
    }

    crumbs.prototype["splitIndexAtCol"] = function (col)
    {
        var si, sl

        sl = 0
        for (var _a_ = si = 0, _b_ = this.split.length; (_a_ <= _b_ ? si < this.split.length : si > this.split.length); (_a_ <= _b_ ? ++si : --si))
        {
            sl += this.split[si].length + 2
            if (sl > col)
            {
                return si
            }
        }
        return this.split.length - 1
    }

    crumbs.prototype["adjustText"] = function ()
    {
        var padding, _110_14_

        if (this.hidden())
        {
            return
        }
        this.path = ((_110_14_=this.path) != null ? _110_14_ : '')
        if (this.path === '')
        {
            this.rounded = ''
            return
        }
        this.split = slash.split(this.path)
        if (!this.dotlessRelative)
        {
            if (!(_k_.in(this.split[0],'~/.')))
            {
                this.split.unshift('/')
            }
        }
        this.root = []
        this.rounded = this.split.join(' ')
        while (this.split.length > 1 && this.rounded.length > this.cells.cols - 2)
        {
            this.root.push(this.split.shift())
            this.rounded = this.split.join(' ')
        }
        this.root = this.root.join('/')
        padding = (this.padLast ? _k_.lpad(this.cells.cols - 2 - this.rounded.length) : '')
        return this.rounded = '' + this.rounded + padding + ''
    }

    crumbs.prototype["set"] = function (path)
    {
        this.cells.rows = 1
        this.path = _k_.trim(path)
        return this.adjustText()
    }

    crumbs.prototype["show"] = function (path)
    {
        return this.set(slash.tilde(path))
    }

    crumbs.prototype["visible"] = function ()
    {
        return this.cells.rows > 0
    }

    crumbs.prototype["onMouse"] = function (event)
    {
        var col, index, path, row, si, _157_26_

        var _a_ = this.cells.posForEvent(event); col = _a_[0]; row = _a_[1]

        crumbs.__super__.onMouse.call(this,event)
        if (!this.hover)
        {
            if ((this.hoverIndex != null))
            {
                delete this.hoverIndex
                return true
            }
            return
        }
        switch (event.type)
        {
            case 'press':
                si = this.splitIndexAtCol(col)
                if ((0 <= this.hoverIndex && this.hoverIndex < this.split.length))
                {
                    path = slash.path.apply(null,this.split.slice(0, typeof si === 'number' ? si+1 : Infinity))
                    path = slash.path(this.root,path)
                    if (!(_k_.in(path[0],'~/')))
                    {
                        path = '/' + path
                    }
                    this.emit('action','click',path,event)
                    delete this.hoverIndex
                }
                return {redraw:true}

            case 'move':
                index = this.splitIndexAtCol(col)
                if (this.hoverIndex !== index)
                {
                    this.hoverIndex = index
                    return {redraw:true}
                }
                break
        }

        return this.hover
    }

    return crumbs
})()

export default crumbs;

File: kakao.app/love/lua/view/base/crumbs.lua
--[[
     0000000  00000000   000   000  00     00  0000000     0000000  
    000       000   000  000   000  000   000  000   000  000       
    000       0000000    000   000  000000000  0000000    0000000   
    000       000   000  000   000  000 0 000  000   000       000  
     0000000  000   000   0000000   000   000  0000000    0000000   

    clickable sequence of path segments with half-circles inbetween 
    used in the status, dircol, browse and searcher file headings
--]]

view = require "view.base.view"


local crumbs = class("crumbs", view)
    


function crumbs:init(name) 
        view.init(self, name)
        
        self.pointerType = 'pointer'
        self.rounded = array()
        
        self:setColor('hover', theme.hover.bg)
        self:setColor('bg', theme.crumbs.bg)
        self:setColor('fg', theme.crumbs.fg)
        self:setColor('empty', theme.editor.bg)
        return self
    end


function crumbs:setColor(key, color) 
        if (key == 'empty') then 
                self:setColor('empty_left', color)
                self:setColor('empty_right', color)
        end
        
        return view.setColor(self, key, color)
    end


function crumbs:layout(x, y, w, h) 
        view.layout(self, x, y, w, h)
        
        return self:adjustText()
    end

-- 0000000    00000000    0000000   000   000  
-- 000   000  000   000  000   000  000 0 000  
-- 000   000  0000000    000000000  000000000  
-- 000   000  000   000  000   000  000   000  
-- 0000000    000   000  000   000  00     00  


function crumbs:draw() 
        if self:hidden() then return end
        if empty(self.rounded) then return end
        
        local colors = array()
        for i in iter(1, #self.split) do 
            if (i == self.hoverIndex) then 
                colors:push(self.color.hover)
            else 
                colors:push(color.darken(self.color.bg, (0.4 + ((0.6 * (i + 1)) / #self.split))))
            end
        end
        
        for x in iter(1, self.rounded:len()) do 
            local si = self:splitIndexAtCol(x)
            local bg = colors[si]
            local ch = self.rounded[x]
            
            if ((ch == '') or (ch == '')) then 
                local fg = bg
                bg = (function () 
    if (si > 1) then 
    return colors[(si - 1)] else 
    return self.color.empty_left
                     end
end)()
                if (x == self.rounded:len()) then 
                    bg = self.color.empty_right
                end
                
                self.cells:set(x, 1, ch, fg, bg)
            else 
                if (si == self.hoverIndex) then 
                    local fg = color.adjustForBackground(self.color.fg, bg)
                elseif (si < (#self.split - 1)) then 
                    local fg = color.darken(self.color.fg, min(1, ((si + 3) / #self.split)))
                else 
                    local fg = color.brighten(self.color.fg)
                end
                
                self.cells:set(x, 1, ch, fg, bg)
            end
        end
        
        return self:render()
    end

--  0000000  00000000   000      000  000000000  000  000   000  0000000    00000000  000   000  
-- 000       000   000  000      000     000     000  0000  000  000   000  000        000 000   
-- 0000000   00000000   000      000     000     000  000 0 000  000   000  0000000     00000    
--      000  000        000      000     000     000  000  0000  000   000  000        000 000   
-- 0000000   000        0000000  000     000     000  000   000  0000000    00000000  000   000  


function crumbs:splitIndexAtCol(col) 
        local sl = 0
        for si = 1, (#self.split + 1)-1 do 
            sl = sl + ((#self.split[si] + 2))
            if (sl >= col) then 
                return si
            end
        end
        
        return #self.split
    end


function crumbs:colsAtSplitIndex(idx) 
        local si = 0
        local ei = 0
        
        print("colsAtSplitIndex " .. tostring(idx) .. "", self.split, type(self.split), #self.split)
        for i in iter(1, idx) do 
            if (i < idx) then 
                si = si + ((#self.split[i] + 2))
            end
            
            ei = ei + ((#self.split[i] + 2))
        end
        
        return array(si, ei)
    end


function crumbs:pathAtSplitIndex(idx) 
        -- log "PATH AT SPLIT INDEX #{idx}" @split
        -- log "PATH AT SPLIT ROOT #{idx}" @root
        local path = slash.path(unpack(self.split:slice(1, idx)))
        -- path = slash.path @root path
        print("PATH AT SPLIT PATH " .. tostring(idx) .. "", path)
        if ((string.sub(path, 1, 1) ~= "~") and (string.sub(path, 1, 1) ~= "/")) then 
            path = '/' .. path
        end
        
        return path
    end

--  0000000   0000000          000  000   000   0000000  000000000  
-- 000   000  000   000        000  000   000  000          000     
-- 000000000  000   000        000  000   000  0000000      000     
-- 000   000  000   000  000   000  000   000       000     000     
-- 000   000  0000000     0000000    0000000   0000000      000     


function crumbs:adjustText() 
        if self:hidden() then return end
        
        self.path = self.path or ''
        
        if (self.path == '') then 
            self.rounded = array()
            return
        end
        
        self.split = slash.split(self.path)
        
        if not self.dotlessRelative then 
            if (kstr.find("~/.", self.split[1]) < 0) then 
                self.split:unshift('/')
            end
        end
        
        self.root = array()
        self.rounded = array('')
        
        while (#self.split > 0) do 
            local s = self.split:shift()
            self.root:push(s)
            self.rounded = self.rounded + (kseg(s))
            self.rounded:push(' ')
        end
        
        self.split = slash.split(self.path)
        self.root = self.root:join('/')
        
        local padding = (function () 
    if self.padLast then 
    return kstr.lpad(((self.cells.cols - 2) - #self.rounded)) else 
    return ''
                  end
end)()
        
        return self.rounded:push('')
    end


function crumbs:set(path) 
        self.cells.rows = 1 -- 𝜏𝖍𝚒𝖘 s⫙ϵ⟅⟅𝖘 ϝ𝚒𝖘𝖍𝛾
        self.path = kstr.trim(path)
        return self:adjustText()
    end


function crumbs:show(path) 
    return self:set(slash.tilde(path))
    end

function crumbs:visible() 
    return (self.cells.rows > 0)
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function crumbs:onMouseLeave(event) 
        if not self.hoverIndex then return end
        local path = self:pathAtSplitIndex(self.hoverIndex)
        local cols = self:colsAtSplitIndex(self.hoverIndex)
        local dict = {index = self.hoverIndex, cols = cols}
        return self:emit('action', 'leave', path, dict)
    end


function crumbs:onMouse(event) 
        view.onMouse(self, event)
        
        if not self.hover then 
            if self.hoverIndex then 
                self.hoverIndex = nil
                return true
            end
            
            return
        end
        
        local col, row = unpack(self:eventPos(event))
        
        if (event.type == 'press') then 
                local si = self:splitIndexAtCol(col)
                if ((1 <= self.hoverIndex) and (self.hoverIndex <= #self.split)) then 
                    local path = self:pathAtSplitIndex(self.hoverIndex)
                    self:emit('action', 'click', path, event)
                end
                
                return true
        elseif (event.type == 'move') then 
                local index = self:splitIndexAtCol(col)
                if (self.hoverIndex ~= index) then 
                    self.hoverIndex = index
                    local cols = self:colsAtSplitIndex(index)
                    local path = self:pathAtSplitIndex(index)
                    local dict = {index = index, cols = cols}
                    self:emit('action', 'enter', path, dict)
                    return true
                end
        end
        
        return self.hover
    end

return crumbs